blockchain/proxy.go

```diff
@@ -717,6 +743,203 @@ func (ep *proxy) IsDataTrieMigrated(ctx context.Context, address sdkCore.Address
+	if filter.FromBlock.HasValue && filter.ToBlock.HasValue {
+		if filter.ToBlock.Value-filter.FromBlock.Value <= MaximumBlocksDelta {
+			return filter.FromBlock.Value, filter.ToBlock.Value, nil
+		}
+		return 0, 0, errors.New("invalid block range: too many blocks to process")
+	}
+
+	// Check if only fromBlock is set
+	if filter.FromBlock.HasValue {
+		toBlock := latestBlock // Set toBlock to latestBlock
+		if toBlock-filter.FromBlock.Value <= MaximumBlocksDelta {
+			return filter.FromBlock.Value, toBlock, nil
+		}
+		return 0, 0, errors.New("invalid block range: too many blocks to process")
+	}
+
+	// Check if only toBlock is set
+	if filter.ToBlock.HasValue {
+		fromBlock := genesisBlock // Set fromBlock to genesisBlock
+		if filter.ToBlock.Value-fromBlock <= MaximumBlocksDelta {
+			return fromBlock, filter.ToBlock.Value, nil
+		}
+		return 0, 0, errors.New("invalid block range: too many blocks to process")
+	}
+
+	return 0, 0, errors.New("no block range specified")
+}
+
+// getBlockNumberByHash retrieves the block number associated with the given block hash
+func (ep *proxy) getBlockNumberByHash(ctx context.Context, shardID uint32, blockHash *[32]byte) (uint64, error) {
+	blockHashStr := hex.EncodeToString(blockHash[:])
+
+	buff, code, err := ep.GetHTTP(ctx, ep.endpointProvider.GetBlockByHash(shardID, blockHashStr))
+	if err != nil || code != http.StatusOK {
+		return 0, createHTTPStatusError(code, err)
+	}
+
+	var response data.BlockResponse
+	if err := json.Unmarshal(buff, &response); err != nil {
+		return 0, err
+	}
+
+	blockNonce := response.Data.Block.Nonce
+
+	// Cache the raw response bytes
+	if ep.filterQueryBlockCacher != nil && len(buff) > 0 {
+		cacheKey := make([]byte, 8)
+		binary.BigEndian.PutUint64(cacheKey, blockNonce)
+		ep.filterQueryBlockCacher.Put(cacheKey, buff, len(buff))
+	}
+
+	return blockNonce, nil
+}
+
+// getLogsFromBlock retrieves logs from a specific block and filters them
+func (ep *proxy) getLogsFromBlock(ctx context.Context, shardID uint32, blockNum uint64, filter *sdkCore.FilterQuery) ([]*transaction.Events, error) {
+	cacheKey := make([]byte, 8)
+	binary.BigEndian.PutUint64(cacheKey, blockNum)
+
+	if ep.filterQueryBlockCacher != nil {
+		cachedResponse, found := ep.filterQueryBlockCacher.Get(cacheKey)
+		if found {
+			cachedBuff, ok := cachedResponse.([]byte)
+			if ok && len(cachedBuff) > 0 {
+				var response data.BlockResponse
+				if err := json.Unmarshal(cachedBuff, &response); err == nil {
+					return extractMatchingEvents(response, filter), nil
+				}
+			}
+		}
+	}
+
+	// Fetch the logs from the network if not found in cache
+	buff, code, err := ep.GetHTTP(ctx, ep.endpointProvider.GetBlockByNonce(shardID, blockNum))
+	if err != nil || code != http.StatusOK {
+		return nil, createHTTPStatusError(code, err)
+	}
+
+	// Cache the raw response bytes
+	if ep.filterQueryBlockCacher != nil && len(buff) > 0 {
+		ep.filterQueryBlockCacher.Put(cacheKey, buff, len(buff))
```

Comment:
same, no nil cacher should be present, the default implementation would do nothing